src/memory.cc

```diff
@@ -236,4 +237,132 @@ AllocatedMemory::~AllocatedMemory()
   }
 }
 
+#ifdef TRITON_ENABLE_GPU
+GrowableMemory::GrowableMemory(
+    size_t byte_size, TRITONSERVER_MemoryType memory_type,
+    int64_t memory_type_id, std::unique_ptr<Allocation>&& allocation,
+    size_t virtual_address_size)
+    : MutableMemory(nullptr, byte_size, memory_type, memory_type_id)
+{
+  allocation_prop_.type = CU_MEM_ALLOCATION_TYPE_PINNED;
+  allocation_prop_.location.type = CU_MEM_LOCATION_TYPE_DEVICE;
+  allocation_prop_.location.id = memory_type_id;
+
+  access_desc_.location = allocation_prop_.location;
+  access_desc_.flags = CU_MEM_ACCESS_FLAGS_PROT_READWRITE;
+  virtual_address_size_ = virtual_address_size;
+  allocation_ = std::move(allocation);
+  virtual_address_offset_ = 0;
+}
+
+Status
+GrowableMemory::Create(
+    std::unique_ptr<GrowableMemory>& growable_memory, size_t byte_size,
+    TRITONSERVER_MemoryType memory_type, int64_t memory_type_id,
+    size_t virtual_address_size)
+{
+  std::unique_ptr<Allocation> allocation =
+      std::make_unique<Allocation>(memory_type_id);
+
+  if (memory_type != TRITONSERVER_MEMORY_GPU) {
+    return Status(
+        Status::Code::INVALID_ARG,
+        std::string("Only TRITONSERVER_MEMORY_GPU is supported for growable "
+                    "memory."));
+  }
+
+  if (byte_size > virtual_address_size) {
+    return Status(
+        Status::Code::INVALID_ARG,
+        std::string("'byte_size' requested for GrowableMemory cannot be smaller"
+                    " than the virtual address size. byte_size: ") +
+            std::to_string(byte_size) +
+            ", virtual_address_size:" + std::to_string(virtual_address_size));
+  }
+
+  RETURN_IF_ERROR(
+      CudaBlockManager::Allocate(byte_size, allocation, memory_type_id));
+
+  void* buffer;
+  RETURN_IF_CUDA_DRIVER_ERR(
```

Comment:
if there is an error - do we clean up allocated memory properly here?

Reply:
Yes, since Allocation is a unique_ptr it will free all the memory upon destruction (i.e., when we return from this function).